packages/auth/src/OAuth.ts

```diff
@@ -1,12 +1,166 @@
+    if (args == null) {
+      throw new Error('Missing OAuth parameters');
+    }
+
+    args.authorization = args.authorization || CARTO_AUTHORIZATION_BASE;
+    args.scopes = args.scopes || '';
+
+    this._emitter = new mitt();
+
+    this._client = new SalteAuth({
+      providers: [
+        new CartoProvider({
+          storage: 'local',
+          level: 'trace',
+          url: args.authorization,
+          redirectUrl: args.redirectURI,
+          clientID: args.clientID,
+          responseType: 'token',
+          scope: args.scopes
+        })
+      ],
+
+      handlers: [
+        new Popup({
+          default: true
+        })
+      ]
+    });
+
+    this._refresher = new Iframe({
+      authorization: args.authorization,
+      redirectURI: refreshUrl || args.redirectURI,
+      clientID: args.clientID,
+      scopes: args.scopes
+    });
+
+    // Read the expires_in, setup timer
+    this._carto = this._client.provider('carto') as CartoProvider;
+
+    if (this._carto.expiresIn !== -1) {
+      this._scheduleRefresh();
+    }
+
+    if (args.scopes) {
+      this._checkScopes(args.scopes);
+    }
+  }
+
+  public login() {
+    return new Promise((resolve, reject) => {
+      const cb = (error?: Error, token?: SalteAuth.EventWrapper) => {
+
+        if (error) {
+          reject(error);
+          this._emitter.emit('error');
+          return;
+        }
+
+        resolve(token);
+        this._emitter.emit('tokenUpdated', token);
+        this._carto.off('login', cb);
+
+        this._scheduleRefresh();
+      };
 
-  constructor(clientId: string) {
-    this._clientId = clientId;
+      this._client.on('login', cb);
+      this._client.login('carto').catch((err) => {
+        reject(err);
+      });
+    });
   }
 
-  public get clientId(): string {
-    return this._clientId;
+  public get client() {
+    return this._client;
+  }
+
+  public get expired() {
+    return this._carto.accessToken && this._carto.accessToken.expired;
+  }
+
+  public get credentials() {
+    if (!this._carto.accessToken) {
+      return null;
+    }
+
+    return new Credentials('roman-carto', this._carto.accessToken.raw);
+  }
+
+  public get token() {
+    if (this._carto.accessToken == null) {
+      return null;
+    }
+
+    return this._carto.accessToken.raw;
+  }
+
+  public get userInfo() {
+    if (this.token === null) {
+      return null;
+    }
+
+    return new UserInfo(this.token, this._carto.userInfo);
+  }
+
+  public clear() {
+    this._carto.reset();
+  }
+
+  public on(type: string, handler: mitt.Handler) {
+    this._emitter.on(type, handler);
+  }
+
+  private _scheduleRefresh() {
+    clearTimeout(this._refreshTimeout);
+    this._refreshTimeout = window.setTimeout(() => {
+      this._refresh();
+    }, this._carto.expiresIn - Date.now() - THRESHOLD);
+  }
+
+  private _refresh() {
+    this._refresher.refresh().then((data) => {
+      // This is nasty, using salte-auth 'internals'
```

Comment:
I guess there's no other way. But maybe it could be hidden after a public 'refresh' method in our CartoProvider

Reply:
Hmm, maybe we should move all of this to CartoProvider and just call $validate from outside, or a proxy method to that.

Reply:
Yeah, I guess we should move some responsability into it, somehow